src/yoroi-wallets/cardano/ShelleyWallet.ts

```diff
@@ -109,227 +116,259 @@ export class ShelleyWallet implements WalletInterface {
+    const {internalChain, externalChain} = addressChains.restore({data, networkId})
 
-    this.transactionCache = await TransactionCache.create(makeStorageWithPrefix(`/wallet/${this.id}/txs`))
+    const wallet = await this.commonCreate({
+      id: walletMeta.id,
+      networkId,
+      storage,
+      internalChain,
+      externalChain,
+
+      implementationId: data.walletImplementationId ?? walletMeta.walletImplementationId, // can be null for versions < 3.0.2
+      accountPubKeyHex: data.publicKeyHex ?? internalChain.publicKey, // can be null for versions < 3.0.2, in which case we can just retrieve from address generator
+      hwDeviceInfo: data.hwDeviceInfo, // hw wallet
+      isReadOnly: data.isReadOnly ?? false, // readonly wallet
+    })
 
-    // initialize address chains
-    const walletConfig = getWalletConfigById(implementationId)
-    this.internalChain = new AddressChain(
-      new AddressGenerator(accountPubKeyHex, 'Internal', implementationId, networkId),
-      walletConfig.DISCOVERY_BLOCK_SIZE,
-      walletConfig.DISCOVERY_GAP_SIZE,
-    )
-    this.externalChain = new AddressChain(
-      new AddressGenerator(accountPubKeyHex, 'External', implementationId, networkId),
-      walletConfig.DISCOVERY_BLOCK_SIZE,
-      walletConfig.DISCOVERY_GAP_SIZE,
-    )
+    wallet.isEasyConfirmationEnabled = data.isEasyConfirmationEnabled
+    wallet.isHW = data.isHW ?? false
+    wallet.state = {lastGeneratedAddressIndex: data.lastGeneratedAddressIndex}
+
+    wallet.integrityCheck()
+
+    return wallet
+  }
+
+  private static commonCreate = async ({
+    id,
+    networkId,
+    implementationId,
+    storage,
+    internalChain,
+    externalChain,
+
+    accountPubKeyHex,
+    hwDeviceInfo, // hw wallet
+    isReadOnly, // readonly wallet
+  }: {
+    accountPubKeyHex: string
+    hwDeviceInfo: HWDeviceInfo | null
+    id: string
+    implementationId: WalletImplementationId
+    networkId: NetworkId
+    storage: typeof storageLegacy
+    internalChain: AddressChain
+    externalChain: AddressChain
+
+    isReadOnly: boolean
+  }) => {
+    const rewardAddressHex = await deriveRewardAddressHex(accountPubKeyHex, networkId)
+    const apiUrl = getCardanoNetworkConfigById(networkId).BACKEND.API_ROOT
+    const utxoManager = await makeUtxoManager(id, apiUrl, storage)
+    const transactionCache = await TransactionCache.create(makeStorageWithPrefix(`/wallet/${id}/txs`))
 
-    this.rewardAddressHex = await deriveRewardAddressHex(accountPubKeyHex, networkId)
+    const wallet = new ShelleyWallet({
+      storage,
+      networkId,
+      id,
+      utxoManager,
+      implementationId,
+      hwDeviceInfo,
+      isReadOnly,
+      provider: undefined,
+      accountPubKeyHex,
+      rewardAddressHex,
+      transactionCache,
+      internalChain,
+      externalChain,
+    })
 
+    await wallet.discoverAddresses()
+    wallet.setupSubscriptions()
```

Comment:
`discoveryAddresses` fire some events dunno about this order. I guess is ok

Reply:
its the same in the current ShelleyWallet version